anomaly-benchmarking/script.whizzml

```diff
@@ -0,0 +1,277 @@
+                                  field-id 
+                                  "summary" 
+                                  "categories"]))))
+    (create-dataset {"origin_dataset" ds
+                     "new_fields" [{"field" (two-class-filter 
+                                             field-id 
+                                             anomalous)
+                                    "name" "binary"}]})))
+
+;; Multi-to-Binary
+
+;; Creating the Evaluation 
+;; (mostly from jao's
+;;  model-per-cluster/create-cluster-models)
+
+;; Samples a dataset with a given rate (between 0 and 1) and getting
+;; the sampled rows (oob = false) or the complementary ones (oob = true),
+;; using a fixed seed.
+
+(define (sample-dataset ds-id rate oob)
+  (create-dataset {"sample_rate" rate
+                   "origin_dataset"  ds-id
+                   "out_of_bag" oob
+                   "seed" "whizzml-example"}))
+
+;; Creates a model with 80% of the input dataset, evaluates it
+;; with the remaining 20%. Returns the evaluation id.
+
+(define (evaluate-model-on-dataset ds-id)
+  (let (training-id (sample-dataset ds-id 0.8 false)
+        test-id (sample-dataset ds-id 0.8 true)
+        model-id (create-model {"dataset" training-id "randomize" true})
+        ev-id (create-and-wait-evaluation {"model" model-id "dataset" test-id}))
+     ev-id))
+
+;; Graphs
+ 
+;; Given an evaluation id, takes the confustion matrix of the model
+;; result and represents it as a complete graph as outlined in
+;; http://www.outlier-analytics.org/odd13kdd/papers/emmott,das,dietterich,fern,wong.pdf,
+;; section 3.2. Each node is a class of the confusion matrix, and each
+;; edge (between nodes j and k) is weighted by summing the value of
+;; cell [j k] and cell [k j].
+
+;; The returned graph is a map with keys "nodes" and "edges". These
+;; are lists of node maps and edge maps, respectively. Node maps have
+;; keys "id" and "name", the class name given in the confusion
+;; matrix. Edge maps have keys "ends", a duple of the node ids the
+;; edge connects, and "weight", its weight calculated from the
+;; confusion matrix.
+
+;;Given a confusion matrix and two nodes, returns the weighted edge.
+
+(define (make-edge matrix two-nodes)
+  (let (node1 (head two-nodes)
+        node2 (last two-nodes)
+        weight12 (nth (nth matrix node1) node2)
+        weight21 (nth (nth matrix node2) node1))
+   (make-map ["ends" "weight"] [[node1 node2] (+ weight12 weight21)]) ))
+
+;; Given a confusion matrix, returns a list of all the properly
+;; weighted edges.
+
+(define (create-edges matrix)
+  (let (count (count matrix))
+    (map (lambda (x) (map (lambda (y) 
+                                  (make-edge matrix [x, y])) 
+                          (range (+ x 1) count))) (range 0 (- count 1)))))
+              
+;; Given a nested list, returns its contents as a single flat
+;; list. Uses recursion in ways bad for big sequences, but works for
```

Comment:
I think Flatline does tail-call optimization, so this should probably be fine even for large sequences.